[Pages] Update Worker site to Pages migration guide (#4233)
cloudflare/cloudflare-docs: content/pages/migrations/migrating-from-workers/index.md

As a prerequisite, you should have a Cloudflare Workers Sites project, created with [Wrangler](https://github.com/cloudflare/wrangler).

> Cloudflare Pages provides built-in defaults for every aspect of serving your site. You can port custom behaviour in your Workers script — such as custom caching logic — to your Cloudflare Pages. Using [Pages Functions](/pages/platform/functions/) which enables you leverage the `/api` filesystem convention serverless functions. You can also migrate your headers and redirects settings to Pages. 

Reviewer: ```suggestion
Cloudflare Pages provides built-in defaults for every aspect of serving your site. You can port custom behaviour in your Workers script — such as custom caching logic — to your Cloudflare Pages project using [Functions](/pages/platform/functions/). This enables an easy to use file-based routing system. You can also migrate your custom headers and redirects to Pages. 
```

## Removing unnecessary code

Workers Sites projects consist of the following pieces:

1.  An application built with a [static site tool](/pages/how-to/) or a static collection of HTML, CSS and JavaScript
2. If using a static site tool, a build directory (called `bucket` in `wrangler.toml`) where the static project builds HTML, CSS, and JS
3. A Workers application for serving that directory

When moving to Cloudflare Pages, remove the Workers application and any associated `wrangler.toml` configuration files or build output. Instead, note and record your **build command** (if you have one), and the `bucket` field, or **build directory**, from the `wrangler.toml` file in your project's directory.

## Migrating Headers and Redirects.

You can migrate your redirects to Pages, by creating a `_redirects` file in your publish directory. Cloudflare currently offers limited support for advanced redirects. More support will be added in the future. Refer to the documentaion for a [list of supported types](/pages/platform/redirects/). 

{{<Aside type="note">}}

> A project is limited to 1,000 static redirects and 100 dynamic redirects, for a combined total of 1,100 redirects. Each redirect declaration has a 1,000-character limit. Malformed definitions are ignored. If there are multiple redirects for the same source path, the topmost redirect is applied. 

Reviewer: ```suggestion
A project is limited to 2,000 static redirects and 100 dynamic redirects, for a combined total of 2,100 redirects. Each redirect declaration has a 1,000-character limit. Malformed definitions are ignored. If there are multiple redirects for the same source path, the topmost redirect is applied. 
```

Author: I'll have to update the limit in the other places redirects are mentioned.

Make sure that static redirects are before dynamic redirects in your `_redirects` file.

{{</Aside>}}

> In addition to an `_redirects` file, Cloudflare Pages currently offers [Bulk Redirects(beta)](/pages/how-to/use-bulk-redirects/), which handles redirects that surpasses the 1,100 redirect rules limit set by Pages.

Reviewer: ```suggestion
In addition to an `_redirects` file, Cloudflare Pages currently offers [Bulk Redirects (beta)](/pages/how-to/use-bulk-redirects/), which handles redirects that surpasses the 1,100 redirect rules limit set by Pages.
```

> Your header files can also be moved into a `_headers` folder in your publish directory. It is important to note that custom headers defined in the `_headers` file are not currently applied to responses from Functions, even if the Function route matches the URL pattern. To learn more about handling headers, refer to [Headers](/pages/platform/headers/).

Reviewer: ```suggestion
Your custom headers can also be moved into a `_headers` file in your publish directory. It is important to note that custom headers defined in the `_headers` file are not currently applied to responses from Functions, even if the Function route matches the URL pattern. To learn more about handling headers, refer to [Headers](/pages/platform/headers/).
```

## Creating a new Pages project

After you have recorded your **build command** and **build directory** in a separate location, remove everything else from your application, and push the new version of your project up to GitHub. Follow the ["Get started" guide](/pages/get-started/) to add your project to Cloudflare Pages, using the **build command** and **build directory** that you saved earlier.

If you choose to use a custom domain for your Pages, you can set it to the same custom domain as your currently deployed Workers application. When Pages finishes the initial deployment of your site, you will need to delete the Workers application to start sending requests to Cloudflare Pages.

> ## Cleaning up your old application and assigning the domain

Reviewer: We don't actually say how to link the custom domain in here. We only say to remove the Worker. This leaves the user with only their pages.dev and a domain without anything being served.

Author: The steps for linking can be found in the getting started guide linked a few lines up. I will make it more obvious.

As a prerequisite, you should have a Cloudflare Workers Sites project, created with [Wrangler](https://github.com/cloudflare/wrangler).

Cloudflare Pages provides built-in defaults for every aspect of serving your site. You can port custom behavior in your Worker — such as custom caching logic — to your Cloudflare Pages project using [Functions](/pages/platform/functions/). This enables an easy-to-use file-based routing system. You can also migrate your custom headers and redirects to Pages. 

You may already have a reasonably complex Worker and/or it would be tedious to splice it up into Pages' file-based routing system. For these cases, Pages offers developers the ability to define a `_worker.js` file in the output directory of your Pages project. 

{{<Aside type="note">}}

When using a `_worker.js` file, the entire `/functions` directory is ignored – this includes its routing and middleware characteristics. Instead, the `_worker.js` file is deployed as is and must be written using the [Module Worker syntax](/workers/runtime-apis/fetch-event/#syntax-module-worker).

{{</Aside>}}

> By Migrating to Cloudflare Pages, you will be able to access features like [preview deployments](/pages/platform/preview-deployments/) and automatic branch deploys with no extra configuration needed.

Reviewer: ```suggestion
By migrating to Cloudflare Pages, you will be able to access features like [preview deployments](/pages/platform/preview-deployments/) and automatic branch deploys with no extra configuration needed.
```

> ## Removing unnecessary code

Reviewer 2: ```suggestion
## Remove unnecessary code
```

Workers Sites projects consist of the following pieces:

> 1.  An application built with a [static site tool](/pages/how-to/) or a static collection of HTML, CSS and JavaScript.

Reviewer 2: ```suggestion
1. An application built with a [static site tool](/pages/how-to/) or a static collection of HTML, CSS and JavaScript.
```

2. If using a static site tool, a build directory (called `bucket` in `wrangler.toml`) where the static project builds HTML, CSS, and JavaScript.

> 3. A Worker application for serving that build directory.

Reviewer 3: I'd suggest mentioning the `workers-site` directory here, since that's the default used by most (all?) Workers Sites templates, and "Worker application" confused me a little when reading.

```suggestion
3. A Worker application for serving that build directory. For most projects, this is likely to be the `workers-site` directory.
```

When moving to Cloudflare Pages, remove the Workers application and any associated `wrangler.toml` configuration files or build output. Instead, note and record your `build` command (if you have one), and the `bucket` field, or build directory, from the `wrangler.toml` file in your project's directory.

## Migrating headers and redirects

You can migrate your redirects to Pages, by creating a `_redirects` file in your publish directory. Cloudflare currently offers limited support for advanced redirects. More support will be added in the future. For a list of support types, refer to the [Redirects documentaion](/pages/platform/redirects/). 

{{<Aside type="note">}}

A project is limited to 2,000 static redirects and 100 dynamic redirects, for a combined total of 2,100 redirects. Each redirect declaration has a 1,000-character limit. Malformed definitions are ignored. If there are multiple redirects for the same source path, the topmost redirect is applied. 

Make sure that static redirects are before dynamic redirects in your `_redirects` file.

{{</Aside>}}

> In addition to an `_redirects` file, Cloudflare Pages currently offers [Bulk Redirects (beta)](/pages/how-to/use-bulk-redirects/), which handles redirects that surpasses the 2,100 redirect rules limit set by Pages.

Reviewer: We don't own the bulk redirects product, probably shouldn't sound like we do
```suggestion
In addition to an `_redirects` file, Cloudflare also offers [Bulk Redirects (beta)](/pages/how-to/use-bulk-redirects/), which handles redirects that surpasses the 2,100 redirect rules limit set by Pages.
```

Your custom headers can also be moved into a `_headers` file in your publish directory. It is important to note that custom headers defined in the `_headers` file are not currently applied to responses from Functions, even if the Function route matches the URL pattern. To learn more about handling headers, refer to [Headers](/pages/platform/headers/).

> ## Creating a new Pages project

Reviewer 2: ```suggestion
## Create a new Pages project
```

After you have recorded your **build command** and **build directory** in a separate location, remove everything else from your application, and push the new version of your project up to GitHub. Follow the ["Get started" guide](/pages/get-started/) to add your project to Cloudflare Pages, using the **build command** and **build directory** that you saved earlier.

If you choose to use a custom domain for your Pages project, you can set it to the same custom domain as your currently deployed Workers application. Follow the steps for [adding a custom domain](/pages/get-started/#adding-a-custom-domain) to your Pages project. 

{{<Aside type="note">}}

> Before you deploy, you will need to delete your Workers application to start sending requests to Cloudflare Pages.

Reviewer 2: ```suggestion
Before you deploy, you will need to delete your old Workers application to start sending requests to Cloudflare Pages.
```

As a prerequisite, you should have a Cloudflare Workers Sites project, created with [Wrangler](https://github.com/cloudflare/wrangler).

> Cloudflare Pages provides built-in defaults for every aspect of serving your site. You can port custom behaviour in your Workers script — such as custom caching logic — to your Cloudflare Pages project using [Functions](/pages/platform/functions/). This enables an easy to use file-based routing system. You can also migrate your custom headers and redirects to Pages. 

Reviewer 2: ```suggestion
Cloudflare Pages provides built-in defaults for every aspect of serving your site. You can port custom behavior in your Worker — such as custom caching logic — to your Cloudflare Pages project using [Functions](/pages/platform/functions/). This enables an easy-to-use file-based routing system. You can also migrate your custom headers and redirects to Pages. 
```

You may already have a reasonably complex Worker and/or it would be tedious to splice it up into Pages' file-based routing system. For these cases, Pages offers developers the ability to define a `_worker.js` file in the output directory of your Pages project. 

{{<Aside type="note">}}

When using a `_worker.js` file, the entire `/functions` directory is ignored – this includes its routing and middleware characteristics. Instead, the `_worker.js` file is deployed as is and must be written using the [Module Worker syntax](/workers/runtime-apis/fetch-event/#syntax-module-worker).

{{</Aside>}}

By Migrating to Cloudflare Pages, you will be able to access features like [preview deployments](/pages/platform/preview-deployments/) and automatic branch deploys with no extra configuration needed.

## Removing unnecessary code

Workers Sites projects consist of the following pieces:

> 1.  An application built with a [static site tool](/pages/how-to/) or a static collection of HTML, CSS and JavaScript

Reviewer 2: ```suggestion
1.  An application built with a [static site tool](/pages/how-to/) or a static collection of HTML, CSS and JavaScript.
```

> 2. If using a static site tool, a build directory (called `bucket` in `wrangler.toml`) where the static project builds HTML, CSS, and JS

Reviewer 2: ```suggestion
2. If using a static site tool, a build directory (called `bucket` in `wrangler.toml`) where the static project builds HTML, CSS, and JavaScript.
```

> 3. A Workers application for serving that directory

Reviewer 2: ```suggestion
3. A Worker application for serving that build directory.
```

> When moving to Cloudflare Pages, remove the Workers application and any associated `wrangler.toml` configuration files or build output. Instead, note and record your **build command** (if you have one), and the `bucket` field, or **build directory**, from the `wrangler.toml` file in your project's directory.

Reviewer 2: ```suggestion
When moving to Cloudflare Pages, remove the Workers application and any associated `wrangler.toml` configuration files or build output. Instead, note and record your build command (if you have one), and the `bucket` field, or build directory, from the `wrangler.toml` file in your project's directory.
```
Are 'build command' and 'build directory' fields in the wrangler.toml file? If so, we should put them in backticks.

Author: Build is a field and build directory refers to the actual value of the bucket 

`[site]
bucket = "./public" # Build directory 
entry-point = "workers-site"`

## Migrating Headers and Redirects.

> You can migrate your redirects to Pages, by creating a `_redirects` file in your publish directory. Cloudflare currently offers limited support for advanced redirects. More support will be added in the future. Refer to the documentaion for a [list of supported types](/pages/platform/redirects/). 

Reviewer 2: ```suggestion
You can migrate your redirects to Pages, by creating a `_redirects` file in your publish directory. Cloudflare currently offers limited support for advanced redirects. More support will be added in the future. For a list of support types, refer to the [Redirects documentaion](/pages/platform/redirects/). 
```

{{<Aside type="note">}}

A project is limited to 2,000 static redirects and 100 dynamic redirects, for a combined total of 2,100 redirects. Each redirect declaration has a 1,000-character limit. Malformed definitions are ignored. If there are multiple redirects for the same source path, the topmost redirect is applied. 

Make sure that static redirects are before dynamic redirects in your `_redirects` file.

{{</Aside>}}

In addition to an `_redirects` file, Cloudflare Pages currently offers [Bulk Redirects (beta)](/pages/how-to/use-bulk-redirects/), which handles redirects that surpasses the 2,100 redirect rules limit set by Pages.

> 

Reviewer 2: ```suggestion
```

Your custom headers can also be moved into a `_headers` file in your publish directory. It is important to note that custom headers defined in the `_headers` file are not currently applied to responses from Functions, even if the Function route matches the URL pattern. To learn more about handling headers, refer to [Headers](/pages/platform/headers/).

## Creating a new Pages project

After you have recorded your **build command** and **build directory** in a separate location, remove everything else from your application, and push the new version of your project up to GitHub. Follow the ["Get started" guide](/pages/get-started/) to add your project to Cloudflare Pages, using the **build command** and **build directory** that you saved earlier.

If you choose to use a custom domain for your Pages project, you can set it to the same custom domain as your currently deployed Workers application. Follow the steps for [adding a custom domain](/pages/get-started/#adding-a-custom-domain) to your Pages project. 

{{<Aside type="note">}}

> Before you deploy, you will need to delete the Workers application to start sending requests to Cloudflare Pages.

Reviewer 2: Is this the user's old Workers Sites project?

Author: It is, but also refers to any functionalities they were handling with Workers.

Reviewer 4: Do we want to recommend that they delete the entire application? Why not just the route? Seems scary.

> Before you deploy, you will need to delete the Workers application to start sending requests to Cloudflare Pages.

Reviewer 2: ```suggestion
Before you deploy, you will need to delete your Workers application to start sending requests to Cloudflare Pages.
```

As a prerequisite, you should have a Cloudflare Workers Sites project, created with [Wrangler](https://github.com/cloudflare/wrangler).

> Cloudflare Pages provides built-in defaults for every aspect of serving your site. You can port custom behavior in your Worker — such as custom caching logic — to your Cloudflare Pages project using [Functions](/pages/platform/functions/). This enables an easy-to-use file-based routing system. You can also migrate your custom headers and redirects to Pages. 

Reviewer 2: ```suggestion
Cloudflare Pages provides built-in defaults for every aspect of serving your site. You can port custom behavior in your Worker — such as custom caching logic — to your Cloudflare Pages project using [Functions](/pages/platform/functions/). This enables an easy-to-use, file-based routing system. You can also migrate your custom headers and redirects to Pages. 
```

You may already have a reasonably complex Worker and/or it would be tedious to splice it up into Pages' file-based routing system. For these cases, Pages offers developers the ability to define a `_worker.js` file in the output directory of your Pages project. 

{{<Aside type="note">}}

When using a `_worker.js` file, the entire `/functions` directory is ignored – this includes its routing and middleware characteristics. Instead, the `_worker.js` file is deployed as is and must be written using the [Module Worker syntax](/workers/runtime-apis/fetch-event/#syntax-module-worker).

{{</Aside>}}

By migrating to Cloudflare Pages, you will be able to access features like [preview deployments](/pages/platform/preview-deployments/) and automatic branch deploys with no extra configuration needed.

> ## Removing unnecessary code

Reviewer 2: ```suggestion
## Remove unnecessary code
```

Workers Sites projects consist of the following pieces:

> 1. An application built with a [static site tool](/pages/how-to/) or a static collection of HTML, CSS and JavaScript.

Reviewer 2: ```suggestion
1. An application built with a [static site tool](/pages/how-to/) or a static collection of HTML, CSS and JavaScript files.
```

> 2. If using a static site tool, a build directory (called `bucket` in `wrangler.toml`) where the static project builds HTML, CSS, and JavaScript.

Reviewer 2: ```suggestion
2. If using a static site tool, a build directory (called `bucket` in `wrangler.toml`) where the static project builds your HTML, CSS, and JavaScript files.
```

3. A Worker application for serving that build directory. For most projects, this is likely to be the `workers-site` directory.

When moving to Cloudflare Pages, remove the Workers application and any associated `wrangler.toml` configuration files or build output. Instead, note and record your `build` command (if you have one), and the `bucket` field, or build directory, from the `wrangler.toml` file in your project's directory.

> ## Migrating headers and redirects

Reviewer 2: ```suggestion
## Migrate headers and redirects
```

{{<Aside type="note">}}

A project is limited to 2,000 static redirects and 100 dynamic redirects, for a combined total of 2,100 redirects. Each redirect declaration has a 1,000-character limit. Malformed definitions are ignored. If there are multiple redirects for the same source path, the topmost redirect is applied. 

Make sure that static redirects are before dynamic redirects in your `_redirects` file.

{{</Aside>}}

In addition to an `_redirects` file, Cloudflare also offers [Bulk Redirects (beta)](/pages/how-to/use-bulk-redirects/), which handles redirects that surpasses the 2,100 redirect rules limit set by Pages.

Your custom headers can also be moved into a `_headers` file in your output directory. It is important to note that custom headers defined in the `_headers` file are not currently applied to responses from Functions, even if the Function route matches the URL pattern. To learn more about handling headers, refer to [Headers](/pages/platform/headers/).

> ## Creating a new Pages project

Reviewer 2: ```suggestion
## Create a new Pages project
```

> ### Using git provider 

Reviewer 2: ```suggestion
### Connect to your git provider
```

> After you have recorded your **build command** and **build directory** in a separate location, remove everything else from your application, and push the new version of your project up to GitHub. Follow the ["Get started" guide](/pages/get-started/) to add your project to Cloudflare Pages, using the **build command** and **build directory** that you saved earlier.

Reviewer 2: ```suggestion
After you have recorded your **build command** and **build directory** in a separate location, remove everything else from your application, and push the new version of your project up to your git provider. Follow the [Get started guide](/pages/get-started/) to add your project to Cloudflare Pages, using the **build command** and **build directory** that you saved earlier.
```
Does this work for GitLab too? If so, changed to git provider to reflect that.

If you choose to use a custom domain for your Pages project, you can set it to the same custom domain as your currently deployed Workers application. Follow the steps for [adding a custom domain](/pages/get-started/#adding-a-custom-domain) to your Pages project. 

{{<Aside type="note">}}

Before you deploy, you will need to delete your old Workers routes to start sending requests to Cloudflare Pages.

{{</Aside>}}

### Using Direct Upload

> If your Workers site has its custom build settings, you can bring your pre-built assets to Pages with [Direct uploads](/pages/platform/direct-upload/). In addition, you can serve your website's assets right to the Cloudflare edge network by either using the [Wrangler CLI](/workers/wrangler/get-started/) or the Drag and drop option. 

Reviewer 2: ```suggestion
If your Workers site has its custom build settings, you can bring your prebuilt assets to Pages with [Direct Uploads](/pages/platform/direct-upload/). In addition, you can serve your website's assets right to the Cloudflare edge network by either using the [Wrangler CLI](/workers/wrangler/get-started/) or the drag and drop option. 
```

Workers Sites projects consist of the following pieces:

1.  An application built with a [static site tool](/pages/how-to/) or a static collection of HTML, CSS and JavaScript.
2. If using a static site tool, a build directory (called `bucket` in `wrangler.toml`) where the static project builds HTML, CSS, and JavaScript.
3. A Worker application for serving that build directory.

When moving to Cloudflare Pages, remove the Workers application and any associated `wrangler.toml` configuration files or build output. Instead, note and record your **build command** (if you have one), and the `bucket` field, or **build directory**, from the `wrangler.toml` file in your project's directory.

> ## Migrating Headers and Redirects.

Reviewer 2: ```suggestion
## Migrate headers and redirects
```

{{<Aside type="note">}}

Before you deploy, you will need to delete your Workers application to start sending requests to Cloudflare Pages.

{{</Aside>}}

> ## Cleaning up your old application and assigning the domain

Reviewer 2: ```suggestion
## Clean up your old application and assign a domain
```
